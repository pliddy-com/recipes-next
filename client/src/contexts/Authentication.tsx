import {
  ReactElement,
  createContext,
  useContext,
  useEffect,
  useState
} from 'react';

import {
  AuthenticationDetails,
  CognitoUser,
  CognitoUserSession
} from 'amazon-cognito-identity-js';

import userPool from 'lib/userPool';

interface SignInProps {
  email: string;
  password: string;
}

interface AuthenticationContextValue {
  getToken(): Promise<null>;
  isAuth: boolean;
  isLoading: boolean;
  signIn({ email, password }: SignInProps): Promise<void>;
  signOut(): void;
}

const AuthenticationContext = createContext<AuthenticationContextValue>(
  {} as AuthenticationContextValue
);

interface AuthenticationProps {
  children: ReactElement | ReactElement[];
}

const AuthenticationProvider = (props: AuthenticationProps) => {
  const [isAuth, setIsAuth] = useState<boolean>(false);
  const [token, setToken] = useState<string | null>();
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const getToken = async () => {
    const user = userPool && userPool.getCurrentUser();

    user &&
      user.getSession((err: Error | null, session: CognitoUserSession) => {
        if (err) {
          return null;
        }
        setToken(session.getIdToken().getJwtToken());
      });

    return null;
  };

  useEffect(() => {
    getToken();
    setIsAuth(token ? true : false);
  }, [token]);

  const signIn = async ({ email, password }: SignInProps) => {
    setIsLoading(true);
    await new Promise((resolve, reject) => {
      const user =
        userPool &&
        new CognitoUser({
          Username: email,
          Pool: userPool
        });

      const authDetails = new AuthenticationDetails({
        Username: email,
        Password: password
      });

      user &&
        user.authenticateUser(authDetails, {
          onSuccess: (result) => {
            setToken(result.getIdToken().getJwtToken());
            resolve(result);
            setIsLoading(false);
          },
          onFailure: (err) => {
            reject(err);
            setIsLoading(false);
          }
          // newPasswordRequired: (data) => {
          //   resolve(data);
          //   setIsLoading(false);
          // }
        });
    });
    setIsLoading(false);
  };

  const signOut = () => {
    const user = userPool && userPool.getCurrentUser();

    user && user.signOut();

    setToken(null);
  };

  return (
    <AuthenticationContext.Provider
      value={{
        getToken,
        isAuth,
        isLoading,
        signIn,
        signOut
      }}
    >
      {props.children}
    </AuthenticationContext.Provider>
  );
};

const useAuthContext = () => useContext(AuthenticationContext);

export { AuthenticationProvider, useAuthContext };
