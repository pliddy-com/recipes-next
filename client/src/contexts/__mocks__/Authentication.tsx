jest.createMockFromModule('../Authentication');

const authContextValues = {
  authLoading: false,
  isAuth: false,
  signIn: jest.fn(),
  signOut: jest.fn(),
  token: 'TOKEN'
};

const useAuthContext = jest.fn().mockImplementation(() => authContextValues);

export { useAuthContext };
