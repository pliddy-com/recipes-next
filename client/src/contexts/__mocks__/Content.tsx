jest.createMockFromModule('../Content');

const cmContextValues = {
  canSave: false,
  editMode: false,
  editLoading: false,
  setCanSave: jest.fn(),
  setEditMode: jest.fn(),
  saveRecipe: jest.fn(),
  setRecipe: jest.fn(),
  setSupressEdit: jest.fn(),
  supressEdit: false,
  toggleEdit: jest.fn()
};

const useContentManagementContext = jest
  .fn()
  .mockImplementation(() => cmContextValues);

export { useContentManagementContext };
