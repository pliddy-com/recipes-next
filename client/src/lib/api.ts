/* istanbul ignore file */

import { queryGraphQLContent } from './gqlClient';
// import * as fs from 'fs';

import {
  filterSlugs,
  filterTagsWithRecipes,
  filterTaxonomyItemsWithRecipes
} from './apiFilters';

import {
  NavMenuDataDocument,
  RecipeIndexDocument,
  RecipeIndexQueryVariables,
  RecipeListDocument,
  RecipeListQueryVariables,
  RecipePageDocument,
  RecipePageQueryVariables,
  RecipeSlugsDocument,
  RecipeSlugsQueryVariables,
  TagSlugsDocument,
  TagSlugsQueryVariables,
  NavMenuDataQuery,
  RecipeSlugsQuery,
  TagSlugsQuery,
  RecipeListQuery,
  LinkedEntriesQuery,
  LinkedEntriesDocument
} from '../types/queries';

import { hasValue } from './utils';

// const ACCESS_TOKEN = process.env.NEXT_PUBLIC_CONTENTFUL_MANAGEMENT_TOKEN ?? '';
// const SPACE_ID = process.env.NEXT_PUBLIC_CONTENTFUL_SPACE_ID ?? '';
// const API_ENDPOINT = process.env.NEXT_PUBLIC_CONTENTFUL_CONTENT_API;

// const client = contentful.createClient({
//   accessToken: ACCESS_TOKEN,
//   space: SPACE_ID
// });

// used to query content for nav menu taxonomy on client layout
export const getNavTaxonomy = async () => {
  const { queryGraphQLContent } = await import('lib/gqlClient');

  const results: NavMenuDataQuery = await queryGraphQLContent(
    NavMenuDataDocument
  );

  const { categories, cuisine, tags } = results;

  return {
    ...(categories && {
      categories: filterTaxonomyItemsWithRecipes({
        taxonomyCollection: categories
      })
    }),
    ...(cuisine && {
      cuisine: filterTaxonomyItemsWithRecipes({ taxonomyCollection: cuisine })
    }),
    ...(tags && {
      tags: filterTagsWithRecipes({ tagCollection: tags })
    })
  };
};

// used to get linked entries for recipe edit

export const getLinkedEntries = async () => {
  const { queryGraphQLContent } = await import('lib/gqlClient');

  const results: LinkedEntriesQuery = await queryGraphQLContent(
    LinkedEntriesDocument
  );

  const { images, tags } = results;

  return {
    ...(images && {
      images: images.items
    }),
    ...(tags && {
      tags: tags.items
    })
  };
};

// used by getStaticProps for recipe page
export const getRecipeSlugs = async (variables: RecipeSlugsQueryVariables) => {
  const results: RecipeSlugsQuery = await queryGraphQLContent(
    RecipeSlugsDocument,
    variables
  );

  const { recipeCollection } = results;

  return recipeCollection && recipeCollection.items
    ? filterSlugs(recipeCollection.items)
    : [];
};

// used by getStaticProps for tag page
export const getTagSlugs = async (variables?: TagSlugsQueryVariables) => {
  const results: TagSlugsQuery = await queryGraphQLContent(
    TagSlugsDocument,
    variables
  );

  const { tagCollection } = results;

  const tags =
    tagCollection &&
    filterTagsWithRecipes({
      tagCollection
    });

  return tags ? filterSlugs(tags) : [];
};

// used to query content for home index page
export const getRecipeIndex = async (variables?: RecipeIndexQueryVariables) => {
  const { recipeCollection } = await queryGraphQLContent(
    RecipeIndexDocument,
    variables
  );

  return recipeCollection ? recipeCollection.items.filter(hasValue) : [];
};

// used to query content for tag & category page
export const getRecipeList = async (variables?: RecipeListQueryVariables) => {
  const { tagCollection } = await queryGraphQLContent(
    RecipeListDocument,
    variables
  );

  return tagCollection ? tagCollection.items.filter(hasValue) : [];
};

// used to query content for standalone recipe page
export const getRecipePage = async (variables?: RecipePageQueryVariables) => {
  const { recipeCollection, categoriesTaxonomy, cuisineTaxonomy } =
    await queryGraphQLContent(RecipePageDocument, variables);

  const categories =
    categoriesTaxonomy?.items.filter(hasValue)[0].childrenCollection?.items;

  const cuisine =
    cuisineTaxonomy?.items.filter(hasValue)[0].childrenCollection?.items;

  const recipe = recipeCollection?.items.filter(hasValue)[0];

  return {
    ...(recipe && { recipe }),
    ...(categories && { categories }),
    ...(cuisine && { cuisine })
  };
};

// used to query content for tag index page
export const getTagIndex = async (variables?: RecipeListQueryVariables) => {
  const results: RecipeListQuery = await queryGraphQLContent(
    RecipeListDocument,
    variables
  );

  const { tagCollection } = results;

  return tagCollection
    ? filterTagsWithRecipes({
        tagCollection
      })
    : [];
};
