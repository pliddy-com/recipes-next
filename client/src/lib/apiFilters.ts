import {
  TagDefaultFragment,
  NavTaxonomyFragment,
  TagTaxonomyFragment,
  RecipeSlugFragment
} from '../types/queries';

import { hasValue } from './utils';

export const filterSlugs = (
  items: (RecipeSlugFragment | TagDefaultFragment | null)[]
) =>
  items &&
  items
    ?.map((child) => (child && child.slug ? child.slug : null))
    .filter(hasValue);

export const filterTagsWithRecipes = ({
  tagCollection
}: {
  tagCollection: TagTaxonomyFragment;
}) =>
  tagCollection &&
  tagCollection?.items
    .filter(hasValue)
    .filter(
      (item) =>
        item?.linkedFrom?.recipeCollection?.total &&
        item?.linkedFrom?.recipeCollection?.total > 0
    );

export const filterTaxonomyItemsWithRecipes = ({
  taxonomyCollection
}: {
  taxonomyCollection: NavTaxonomyFragment;
}) => {
  return taxonomyCollection?.items?.[0]?.childrenCollection?.items?.filter(
    (item) =>
      item &&
      (('linkedFrom' in item &&
        item.linkedFrom &&
        'recipeCollection' in item.linkedFrom && // is a tag
        item?.linkedFrom?.recipeCollection?.total &&
        item?.linkedFrom?.recipeCollection?.total > 0) ||
        ('childrenCollection' in item && // is a taxonomy
          item?.childrenCollection?.total &&
          item?.childrenCollection?.total > 0))
  );
};

// TODO: update to filter out childrenCollection.items.linkedFrom.recipeCollection.total <== 0
