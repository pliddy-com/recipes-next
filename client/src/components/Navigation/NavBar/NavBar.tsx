import { useRouter } from 'next/router';

import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Toolbar from '@mui/material/Toolbar';

import EditIcon from '@mui/icons-material/Edit';
import LanguageIcon from '@mui/icons-material/Language';
import MenuIcon from '@mui/icons-material/Menu';
import TagIcon from '@mui/icons-material/Tag';

import LogoButton from 'components/Navigation/Buttons/LogoButton/LogoButton';
import MobileNavControl from 'components/Navigation/MobileNavControl/MobileNavControl';
import NavMenuControl from 'components/Navigation/NavMenuControl/NavMenuControl';
import SearchButton from 'components/Navigation/Buttons/SearchButton/SearchButton';
import SignInControl from '../../SignIn/SignInControl/SignInControl';

import { useAuthContext } from 'contexts/Authentication';

import { TaxonomyChildrenItem } from 'types/queries';

export interface NavDataProps {
  categories: TaxonomyChildrenItem[];
  cuisine: TaxonomyChildrenItem[];
  tags: TaxonomyChildrenItem[];
}

export interface NavBarProps {
  nav?: NavDataProps;
}

const NavBar = ({ nav }: NavBarProps) => {
  const { categories, cuisine, tags } = nav ?? {};
  const { isAuth, toggleEdit } = useAuthContext();

  const { asPath } = useRouter();

  const isRecipe = asPath.startsWith('/recipes/');

  return (
    <AppBar component="nav" data-testid="navbar">
      <Toolbar className="navbar">
        <LogoButton />

        {nav ? (
          <>
            <Box className="mobile-nav">
              <SignInControl hideLabel={true} />
              <MobileNavControl ariaLabel="open navigation menu" nav={nav} />
            </Box>

            <Box className="desktop-nav">
              {categories && (
                <NavMenuControl
                  ariaLabel="open categories menu"
                  label="Categories"
                  icon={<MenuIcon />}
                  data-testid="categories-menu"
                  featuredLabel="All Recipes"
                  featuredUrl="/"
                  id="categories"
                  nav={categories}
                  root="tags"
                />
              )}

              {cuisine && (
                <NavMenuControl
                  ariaLabel="open cuisine menu"
                  label="Cuisine"
                  icon={<LanguageIcon />}
                  data-testid="cuisine-menu"
                  id="cuisine"
                  nav={cuisine}
                  root="tags"
                />
              )}

              {tags && (
                <NavMenuControl
                  ariaLabel="open tags menu"
                  label="Tags"
                  icon={<TagIcon />}
                  data-testid="tags-menu"
                  featuredLabel="All Tags"
                  featuredUrl="/tags"
                  id="tags"
                  nav={tags}
                  root="tags"
                />
              )}

              <SearchButton />
              <SignInControl />
            </Box>
          </>
        ) : null}
      </Toolbar>
      {isAuth && isRecipe ? (
        <Toolbar className="userBar">
          <Button
            className="edit"
            onClick={toggleEdit}
            size="small"
            startIcon={<EditIcon />}
            variant="text"
          >
            Edit
          </Button>
        </Toolbar>
      ) : null}
    </AppBar>
  );
};

export default NavBar;
