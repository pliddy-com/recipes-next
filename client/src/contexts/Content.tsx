/* istanbul ignore file */

import {
  Dispatch,
  ReactElement,
  SetStateAction,
  SyntheticEvent,
  createContext,
  useContext,
  useEffect,
  useState
} from 'react';

import { useRouter } from 'next/router';
import { useAuthContext } from './Authentication';
import { IRecipeChangeSet } from 'types/content';

interface ICMContext {
  canSave: boolean;
  editMode: boolean;
  editLoading: boolean;
  setCanSave: Dispatch<SetStateAction<boolean>>;
  setEditMode: Dispatch<SetStateAction<boolean>>;
  saveRecipe(event: SyntheticEvent): void;
  setRecipe: Dispatch<SetStateAction<IRecipeChangeSet | undefined>>;
  supressEdit: boolean;
  setSupressEdit: Dispatch<SetStateAction<boolean>>;
  toggleEdit(): void;
  uploadImage(file: File): void;
}

interface ContentManagementProps {
  children: ReactElement | ReactElement[];
}

const recipeApiUrl = `${process.env.NEXT_PUBLIC_AWS_API}/recipes`;
const assetsApiUrl = `${process.env.NEXT_PUBLIC_AWS_API}/assets`;

const ContentManagementContext = createContext<ICMContext>({} as ICMContext);

const ContentManagementProvider = (props: ContentManagementProps) => {
  const [editMode, setEditMode] = useState<boolean>(false);
  const [recipe, setRecipe] = useState<IRecipeChangeSet | undefined>();
  const [editLoading, setEditLoading] = useState<boolean>(false);
  const [canSave, setCanSave] = useState<boolean>(false);
  const [supressEdit, setSupressEdit] = useState<boolean>(false);

  const { isAuth, token } = useAuthContext();
  const { asPath, push } = useRouter();

  useEffect(() => {
    if (!isAuth) setEditMode(false);
  }, [isAuth]);

  const toggleEdit = () => {
    setEditMode(!editMode);
  };

  const updateEntry = async ({ recipe }: { recipe: IRecipeChangeSet }) => {
    if (!isAuth || !token) {
      throw new Error('User is not authenticated.');
    }

    if (!canSave) {
      throw new Error('Content has not changed.');
    }

    const restApi = `${recipeApiUrl}/${recipe.id}`;

    try {
      const response = await fetch(restApi, {
        method: 'PUT',
        headers: {
          Authorization: `${token}`
        },
        body: JSON.stringify(recipe)
      });

      return response.json();
    } catch (e) {
      console.error('Could not save changes to the recipe.');
    }
  };

  const saveRecipe = async (event: SyntheticEvent) => {
    setEditLoading(true);
    event.preventDefault();

    if (!isAuth || !token) {
      throw new Error('User is not authenticated');
    }

    if (recipe) {
      try {
        recipe && (await updateEntry({ recipe }));

        const page = asPath.split('/').slice(-1)[0];
        if (recipe && page !== recipe.slug) push('/');

        setEditMode(false);
        setEditLoading(false);
      } catch (e) {
        // TODO: handle error in UI
        console.error(e);
      }
    } else {
      const err = new Error('No recipe payload to save.');
      throw err;
    }
  };

  // upload media asset to Contentful

  // MOVE TO AWS API
  // UPLOAD IMAGE TO S3 USING SECURE URL
  // SEND SECURE DOWNLOAD URL TO CONTENTFUL WITH CREATE ASSET REQUEST

  const bufferToText = (buffer: ArrayBufferLike) => {
    const bufferByteLength = buffer.byteLength;
    const bufferUint8Array = new Uint8Array(buffer, 0, bufferByteLength);

    return new TextDecoder().decode(bufferUint8Array);
  };

  const uploadImage = async (file: File) => {
    try {
      if (!isAuth || !token) {
        throw new Error('User is not authenticated.');
      }

      console.log({ file });

      // const { name, type } = file;

      const buffer = await file.arrayBuffer();
      const body = bufferToText(buffer);

      console.log('body', body);

      try {
        const response = await fetch(assetsApiUrl, {
          method: 'POST',
          headers: {
            Authorization: `${token}`
          },
          body: JSON.stringify(body)
        });

        return response.json();
      } catch (e) {
        console.error('Could not upload the image.');
      }

      // const upload =
      //   buffer &&
      //   (await client
      //     .getSpace(SPACE_ID)
      //     .then((space) => space.getEnvironment('master'))
      //     .then((environment) =>
      //       environment.createAssetFromFiles({
      //         fields: {
      //           title: {
      //             'en-US': name
      //           },
      //           description: {
      //             'en-US': 'Asset description'
      //           },
      //           file: {
      //             'en-US': {
      //               contentType: type,
      //               fileName: name,
      //               file: buffer
      //             }
      //           }
      //         }
      //       })
      //     )
      //     .then((asset) => {
      //       console.log('processing upload');
      //       return asset.processForAllLocales({ processingCheckWait: 2000 });
      //     })
      //     .then((asset) => {
      //       console.log('publishing upload');
      //       return asset.publish();
      //     })
      //     .then((asset) => {
      //       console.log({ asset });
      //       return asset;
      //     }));

      // console.log('Upload ', upload);
    } catch (e) {
      console.error(e);
    }
  };

  return (
    <ContentManagementContext.Provider
      value={{
        canSave,
        editLoading,
        editMode,
        saveRecipe,
        setCanSave,
        setEditMode,
        setRecipe,
        setSupressEdit,
        supressEdit,
        toggleEdit,
        uploadImage
      }}
    >
      {props.children}
    </ContentManagementContext.Provider>
  );
};

const useContentManagementContext = () => useContext(ContentManagementContext);

export { ContentManagementProvider, useContentManagementContext };
