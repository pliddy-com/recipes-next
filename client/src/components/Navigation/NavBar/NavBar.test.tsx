// import testing-library methods
import { render } from '@testing-library/react';

// add custom jest matchers from jest-dom
import '@testing-library/jest-dom';

// import the component to test
import NavBar, { NavDataProps } from './NavBar';

import * as api from 'lib/api';
import * as AuthContext from 'contexts/Authentication';
import * as ContentManagementContext from 'contexts/Content';

jest.mock('lib/api');
jest.mock('contexts/Authentication');
jest.mock('contexts/Content');

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({
    asPath: '/recipes/',
    push: jest.fn(),
    back: jest.fn(),
    events: {
      on: jest.fn(),
      off: jest.fn()
    }
  })
}));

const authContextValues = {
  authLoading: false,
  isAuth: false,
  signIn: jest.fn(),
  signOut: jest.fn(),
  token: 'TOKEN'
};

const cmContextValues = {
  canSave: false,
  editMode: false,
  editLoading: false,
  setCanSave: jest.fn(),
  saveRecipe: jest.fn(),
  setRecipe: jest.fn(),
  setSupressEdit: jest.fn(),
  supressEdit: false,
  toggleEdit: jest.fn()
};

describe('NavBar', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn() // Deprecated
      }))
    });
  });

  describe('when there is a properly structured nav property', () => {
    it('it renders a nav bar', async () => {
      const authSpy = jest
        .spyOn(AuthContext, 'useAuthContext')
        .mockImplementation(() => authContextValues);

      const cmSpy = jest
        .spyOn(ContentManagementContext, 'useContentManagementContext')
        .mockImplementation(() => cmContextValues);
      const nav = await api.getNavTaxonomy();

      const { asFragment, queryByRole, queryByTestId } = render(
        <NavBar nav={nav as NavDataProps} />
      );

      expect(authSpy).toBeCalled();
      expect(cmSpy).toBeCalled();

      const navBar = queryByTestId('navbar');

      // assert that component has been rendered
      expect(navBar).toBeInTheDocument();

      // assert that buttons have been rendered
      const logoButton = queryByRole('button', {
        name: 'home'
      });

      expect(logoButton).toBeDefined();
      // assert that component has correct href
      expect(logoButton).toHaveAttribute('href', '/');

      const searchButton = queryByRole('button', {
        name: 'search'
      });

      expect(searchButton).toBeDefined();
      expect(searchButton).toHaveAttribute('href', '/search');

      // assert that the component matches the existing snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there is no nav property', () => {
    it('it does not render nav items', () => {
      const { queryByRole } = render(<NavBar nav={undefined} />);

      // assert that there is no NavMenu
      const mobileNav = queryByRole('button', { name: 'open navigation menu' });
      expect(mobileNav).toBeNull();

      const desktopNav = queryByRole('button', {
        name: 'open categories menu'
      });
      expect(desktopNav).toBeNull();
    });
  });

  describe('when isAuth is true', () => {
    it('it shows the user toolbar', async () => {
      const testAuth = { ...authContextValues, isAuth: true };

      const authSpy = jest
        .spyOn(AuthContext, 'useAuthContext')
        .mockImplementation(() => testAuth);

      const cmSpy = jest
        .spyOn(ContentManagementContext, 'useContentManagementContext')
        .mockImplementation(() => cmContextValues);

      const nav = await api.getNavTaxonomy();

      const { asFragment } = render(<NavBar nav={nav as NavDataProps} />);

      expect(authSpy).toBeCalled();
      expect(cmSpy).toBeCalled();

      // assert that the component matches the existing snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when editMode is true', () => {
    it('it shows the cancel button', async () => {
      const testAuthContext = { ...authContextValues, isAuth: true };
      const testContentManagementContext = {
        ...cmContextValues,
        editMode: true
      };

      const authSpy = jest
        .spyOn(AuthContext, 'useAuthContext')
        .mockImplementation(() => testAuthContext);

      const cmSpy = jest
        .spyOn(ContentManagementContext, 'useContentManagementContext')
        .mockImplementation(() => testContentManagementContext);

      const nav = await api.getNavTaxonomy();

      const { asFragment } = render(<NavBar nav={nav as NavDataProps} />);

      expect(authSpy).toBeCalled();
      expect(cmSpy).toBeCalled();

      // assert that the component matches the existing snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when canSave is true', () => {
    it('it activates the save button', async () => {
      const testAuthContext = { ...authContextValues, isAuth: true };
      const testContentManagementContext = {
        ...cmContextValues,
        canSave: true,
        editMode: true
      };

      const authSpy = jest
        .spyOn(AuthContext, 'useAuthContext')
        .mockImplementation(() => testAuthContext);

      const cmSpy = jest
        .spyOn(ContentManagementContext, 'useContentManagementContext')
        .mockImplementation(() => testContentManagementContext);

      const nav = await api.getNavTaxonomy();

      const { asFragment } = render(<NavBar nav={nav as NavDataProps} />);

      expect(authSpy).toBeCalled();
      expect(cmSpy).toBeCalled();

      // assert that the component matches the existing snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
