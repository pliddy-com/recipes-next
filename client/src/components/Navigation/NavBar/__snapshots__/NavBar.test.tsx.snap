// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar when there is a properly structured nav property it renders a nav bar 1`] = `
<div>
  <nav
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
    data-testid="navbar"
    style=""
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular menu css-hyum1k-MuiToolbar-root"
    >
      <a
        aria-label="home"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge logo css-kpm0tx-MuiButtonBase-root-MuiButton-root"
        href="/"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
        >
          <div
            aria-label="logo"
            class="MuiBox-root css-0"
            data-testid="logo"
            role="graphics-symbol"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12.047 8c-4.906 0-8.248.82-8.715 1.012-.467.19-1.116.3-1.57.597-.448.447-.929.384-1.323.393-.179.002-.365.069-.418.154-.157.26.59 1.072.989 1.074.692.006.802.248 1.074 2.372.448 3.498.978 5.418 1.652 5.97.126.104.808.384 1.518.623 1.468.495 1.491.533 1.527 2.623.016.916.078 1.182.276 1.182.29 0 .37-.308.523-1.963l.107-1.172.547.123c.709.16 6.916.161 7.621 0l.543-.12.122 1.564c.107 1.377.16 1.568.447 1.568.294 0 .326-.144.326-1.453 0-.9.082-1.56.215-1.73.118-.152.756-.452 1.418-.663.684-.22 1.369-.57 1.586-.81.555-.617 1.02-2.338 1.424-5.264.201-1.461.444-2.634.564-2.71.114-.076.36-.137.541-.14.435-.003.959-.498.959-.904 0-.227-.127-.323-.43-.324-.39-.008-.708.06-1.258-.406-.452-.416-2.265-.985-3.908-1.242C17.144 8.188 15.8 8 12.047 8Zm.312.705c2.663.015 5.259.244 6.946.682 1.332.346 1.729.632 1.466 1.056-.267.434-1.471.808-3.457 1.073-5.785.77-14.1.051-14.1-1.221 0-.416.566-.728 1.8-.99 1.954-.417 4.683-.615 7.345-.6Z"
                style="color: rgb(255, 255, 255);"
              />
            </svg>
          </div>
        </span>
        Patrick's Recipes
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </a>
      <button
        aria-label="open categories menu"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge menu-button css-kpm0tx-MuiButtonBase-root-MuiButton-root"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="MenuIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        Categories
        <span
          class="dropdown-icon"
        >
          ▼
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="open cuisine menu"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge menu-button css-kpm0tx-MuiButtonBase-root-MuiButton-root"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="LanguageIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm6.93 6h-2.95c-.32-1.25-.78-2.45-1.38-3.56 1.84.63 3.37 1.91 4.33 3.56zM12 4.04c.83 1.2 1.48 2.53 1.91 3.96h-3.82c.43-1.43 1.08-2.76 1.91-3.96zM4.26 14C4.1 13.36 4 12.69 4 12s.1-1.36.26-2h3.38c-.08.66-.14 1.32-.14 2 0 .68.06 1.34.14 2H4.26zm.82 2h2.95c.32 1.25.78 2.45 1.38 3.56-1.84-.63-3.37-1.9-4.33-3.56zm2.95-8H5.08c.96-1.66 2.49-2.93 4.33-3.56C8.81 5.55 8.35 6.75 8.03 8zM12 19.96c-.83-1.2-1.48-2.53-1.91-3.96h3.82c-.43 1.43-1.08 2.76-1.91 3.96zM14.34 14H9.66c-.09-.66-.16-1.32-.16-2 0-.68.07-1.35.16-2h4.68c.09.65.16 1.32.16 2 0 .68-.07 1.34-.16 2zm.25 5.56c.6-1.11 1.06-2.31 1.38-3.56h2.95c-.96 1.65-2.49 2.93-4.33 3.56zM16.36 14c.08-.66.14-1.32.14-2 0-.68-.06-1.34-.14-2h3.38c.16.64.26 1.31.26 2s-.1 1.36-.26 2h-3.38z"
            />
          </svg>
        </span>
        Cuisine
        <span
          class="dropdown-icon"
        >
          ▼
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="open tags menu"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge menu-button css-kpm0tx-MuiButtonBase-root-MuiButton-root"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="TagIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 10V8h-4V4h-2v4h-4V4H8v4H4v2h4v4H4v2h4v4h2v-4h4v4h2v-4h4v-2h-4v-4h4zm-6 4h-4v-4h4v4z"
            />
          </svg>
        </span>
        Tags
        <span
          class="dropdown-icon"
        >
          ▼
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <a
        aria-label="search"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge menu-button css-kpm0tx-MuiButtonBase-root-MuiButton-root"
        href="/search"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SearchIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </span>
        Search
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </a>
    </div>
  </nav>
</div>
`;
